File: backend/app/services/security_event_store.py
import uuid
from collections import deque
from datetime import datetime, timezone
from typing import List, Dict, Any


class SecurityEventStore:
    """
    In-memory storage for logging historical prompt validation events.
    Thread-safe and capped at the latest 50 events.
    """

    def __init__(self, max_size: int = 50):
        self._events = deque(maxlen=max_size)

    def add_event(self, allowed: bool, label: str, source: str, summary_reason: str) -> Dict[str, Any]:
        """
        Constructs and appends a new security event to the store.
        """
        event = {
            "id": str(uuid.uuid4()),
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "allowed": allowed,
            "label": label,
            "source": source,
            "summary_reason": summary_reason,
        }
        self._events.append(event)
        return event

    def get_events(self) -> List[Dict[str, Any]]:
        """
        Returns recent events with newest first.
        """
        return list(reversed(self._events))


# Export a singleton instance
security_event_store = SecurityEventStore()

File: backend-node/src/services/securityEventStore.ts
import { randomUUID } from 'node:crypto'

export interface SecurityEvent {
  id: string
  timestamp: string
  allowed: boolean
  label: string
  source: string
  summary_reason: string
}

const MAX_SIZE = 50

class SecurityEventStore {
  private events: SecurityEvent[] = []

  addEvent(allowed: boolean, label: string, source: string, summaryReason: string): SecurityEvent {
    const event: SecurityEvent = {
      id: randomUUID(),
      timestamp: new Date().toISOString(),
      allowed,
      label,
      source,
      summary_reason: summaryReason,
    }
    this.events.push(event)
    if (this.events.length > MAX_SIZE) {
      this.events.shift()
    }
    return event
  }

  getEvents(): SecurityEvent[] {
    return [...this.events].reverse()
  }
}

export const securityEventStore = new SecurityEventStore()
